src/shared/codelens/codeLensUtils.ts

```diff
@@ -208,100 +292,141 @@ export async function pickAddSamDebugConfiguration(
 }
 
-export async function makePythonCodeLensProvider(): Promise<vscode.CodeLensProvider> {
+export interface OverridableCodeLensProvider extends vscode.CodeLensProvider {
+    provideCodeLenses: (
+        document: vscode.TextDocument,
+        token: vscode.CancellationToken,
+        forceProvide?: boolean
+    ) => Promise<vscode.CodeLens[]>
+}
+
+export async function makePythonCodeLensProvider(
+    configuration: SettingsConfiguration
+): Promise<OverridableCodeLensProvider> {
     return {
         // CodeLensProvider
         provideCodeLenses: async (
             document: vscode.TextDocument,
-            token: vscode.CancellationToken
+            token: vscode.CancellationToken,
+            forceProvide?: boolean
         ): Promise<vscode.CodeLens[]> => {
-            // Try to activate the Python Extension before requesting symbols from a python file
-            await pythonDebug.activatePythonExtensionIfInstalled()
-            if (token.isCancellationRequested) {
-                return []
+            if (forceProvide || configuration.readSetting<boolean>(STATE_NAME_ENABLE_CODELENSES)) {
+                // Try to activate the Python Extension before requesting symbols from a python file
+                await pythonDebug.activatePythonExtensionIfInstalled()
+                if (token.isCancellationRequested) {
+                    return []
+                }
+
+                const handlers: LambdaHandlerCandidate[] = await pythonCodelens.getLambdaHandlerCandidates(document.uri)
+                return makeCodeLenses({
+                    document,
+                    handlers,
+                    token,
+                    runtimeFamily: RuntimeFamily.Python,
+                })
             }
 
-            const handlers: LambdaHandlerCandidate[] = await pythonCodelens.getLambdaHandlerCandidates(document.uri)
-            return makeCodeLenses({
-                document,
-                handlers,
-                token,
-                runtimeFamily: RuntimeFamily.Python,
-            })
+            return []
         },
     }
 }
 
-export async function makeCSharpCodeLensProvider(): Promise<vscode.CodeLensProvider> {
+export async function makeCSharpCodeLensProvider(
+    configuration: SettingsConfiguration
+): Promise<OverridableCodeLensProvider> {
     return {
         provideCodeLenses: async (
             document: vscode.TextDocument,
-            token: vscode.CancellationToken
+            token: vscode.CancellationToken,
+            forceProvide?: boolean
         ): Promise<vscode.CodeLens[]> => {
-            const handlers: LambdaHandlerCandidate[] = await csharpCodelens.getLambdaHandlerCandidates(document)
-            return makeCodeLenses({
-                document,
-                handlers,
-                token,
-                runtimeFamily: RuntimeFamily.DotNetCore,
-            })
+            if (forceProvide || configuration.readSetting<boolean>(STATE_NAME_ENABLE_CODELENSES)) {
```

Comment:
in all of these `makeXCodeLensProvider` cases, returning early has these advantages:

1.  avoids code churn
2. avoids unnecessary nesting/indentation by returning early
3. easier to read

```suggestion
            if (!forceProvide && !configuration.readSetting<boolean>(STATE_NAME_ENABLE_CODELENSES)) {
                return []
            }
```